src/main/java/org/asamk/signal/dbus/DbusSignalImpl.java

```diff
@@ -60,6 +64,54 @@ public String getObjectPath() {
         return objectPath;
     }
 
+    @Override
+    public void addDevice(String uri) {
+        try {
+            m.addDeviceLink(new URI(uri));
+        } catch (IOException | InvalidKeyException e) {
+            throw new Error.Failure(e.getClass().getSimpleName() + " Add device link failed. " + e.getMessage());
+        } catch (URISyntaxException e) {
+            throw new Error.Failure(e.getClass().getSimpleName() + " Device link uri has invalid format: " + e.getMessage());
+        }
+    }
+
+    @Override
+    public void removeDevice(int deviceId) {
+        try {
+            m.removeLinkedDevices(deviceId);
+        } catch (IOException e) {
+            throw new Error.Failure(e.getClass().getSimpleName() + ": Error while removing device: " + e.getMessage());
+        }
+    }
+
+    @Override
+    public List<String> listDevices() {
+        List<Device> devices;
+        List<String> results = new ArrayList<String>();
+
+        try {
+            devices = m.getLinkedDevices();
+        } catch (IOException | Error.Failure e) {
+            throw new Error.Failure("Failed to get linked devices: " + e.getMessage());
+        }
+
+        for (var d : devices) {
+            var name = d.getName();
+            if (name == null) {name = "null";}
+            results.add(name);
+        }
+        return results;
```

Comment:
You could replace this for loop with a stream.
```suggestion
        return devices.stream()
            .map(d -> d.getName() == null ? "" : d.getName())
            .collect(Collectors.toList());
```

Reply:
syntax still baffles me, but hey it works :)